ground/src/main/java/com/google/android/ground/persistence/remote/firestore/schema/LoiConverter.kt

```diff
@@ -37,52 +34,48 @@ object LoiConverter {
     const val POLYGON_TYPE = "Polygon"
     const val GEOMETRY_COORDINATES = "coordinates"
     const val GEOMETRY = "geometry"
+    private val geometryFactory = GeometryFactory()
 
     @JvmStatic
     @Throws(DataStoreException::class, NotImplementedError::class)
     fun toLoi(survey: Survey, doc: DocumentSnapshot): LocationOfInterest {
         val loiId = doc.id
-        val loiDoc = checkNotNull(doc.toObject(LoiDocument::class.java), "LOI data")
-        val geometryMap = checkNotNull(loiDoc.geometry, "geometry")
+        val loiDoc =
+            DataStoreException.checkNotNull(doc.toObject(LoiDocument::class.java), "LOI data")
+        val geometryMap = DataStoreException.checkNotNull(loiDoc.geometry, "geometry")
         // TODO: Return `Result` instead of throwing exception.
         val geometry = GeometryConverter.fromFirestoreMap(geometryMap).getOrThrow()
 
         // As an interim solution, we map geometries to existing LOI types.
         // TODO: Get rid of LOI subclasses and just use Geometry on LOI class.
         when (geometry.geometryType) {
-            "Point" -> {
-                val builder = PointOfInterest.newBuilder()
-                builder.setPoint(
-                    Point.newBuilder().setLatitude(geometry.coordinate.x)
-                        .setLongitude(geometry.coordinate.y).build()
+            "Point" ->
+                return fillLocationOfInterest(
+                    survey,
+                    loiId,
+                    loiDoc,
+                    geometryFactory.createPoint(geometry.coordinate)
                 )
-                fillLocationOfInterest(builder, survey, loiId, loiDoc)
-                return builder.build()
-            }
-            "Polygon" -> {
-                val builder = AreaOfInterest.newBuilder()
-                builder.setVertices(geometry.coordinates.map {
-                    Point.newBuilder().setLatitude(it.x).setLongitude(it.y).build()
-                }.toImmutableList())
-                fillLocationOfInterest(builder, survey, loiId, loiDoc)
-                return builder.build()
-            }
+            "Polygon" -> return fillLocationOfInterest(
+                survey,
+                loiId,
+                loiDoc,
+                geometryFactory.createPolygon(geometry.coordinates)
+            )
```

Comment:
Instead of repeating the constructor code, can we use `.map()` on line 47 to convert to the appropriate geometry?

Reply:
great idea!